tensorflow/contrib/correlation_cost/kernels/correlation_cost_op.cc

```diff
@@ -0,0 +1,369 @@
+            }
+          }
+        }
+      }
+    }
+    return Status::OK();
+  }
+};
+
+}  // end namespace functor
+
+template <typename Device, typename T>
+class CorrelationCostOp : public OpKernel {
+ public:
+  explicit CorrelationCostOp(OpKernelConstruction* context)
+      : OpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("kernel_size", &kernel_size));
+    OP_REQUIRES_OK(context,
+                   context->GetAttr("max_displacement", &max_displacement));
+    OP_REQUIRES_OK(context, context->GetAttr("stride_1", &stride_1));
+    OP_REQUIRES_OK(context, context->GetAttr("stride_2", &stride_2));
+    OP_REQUIRES_OK(context, context->GetAttr("pad", &pad));
+    string data_format;
+    OP_REQUIRES_OK(context, context->GetAttr("data_format", &data_format));
+    OP_REQUIRES(context, FormatFromString(data_format, &data_format_),
+                errors::InvalidArgument("Invalid data format"));
+    OP_REQUIRES(context, kernel_size % 2 != 0,
+                errors::InvalidArgument("kernel_size must be odd"));
+  }
+
+  void Compute(OpKernelContext* context) override {
+    const Tensor& input_a_t = context->input(0);
+    const Tensor& input_b_t = context->input(1);
+
+    // we didn't check the batch-dimension during "SetShapeFn"
+    OP_REQUIRES(context, input_a_t.shape() == input_b_t.shape(),
+                errors::InvalidArgument("Input shapes have to be the same"));
+
+    const int32 N = GetTensorDim(input_a_t, data_format_, 'N');
+    const int32 C = GetTensorDim(input_a_t, data_format_, 'C');
+    const int32 H = GetTensorDim(input_a_t, data_format_, 'H');
+    const int32 W = GetTensorDim(input_a_t, data_format_, 'W');
+
+    // output channels are d**2 where, d = 2r + 1
+    const int32 r = max_displacement / stride_2;
+    const int32 d = 2 * r + 1;
+    const int32 border = max_displacement + (kernel_size - 1) / 2;
+
+    const int32 Cout = d * d;
+    // for spatial dimensions, we pad the inputs
+    const int32 Hout =
+        static_cast<int>(ceil(static_cast<float>(((H + 2 * pad) - border * 2)) /
+                              static_cast<float>(stride_1)));
+    const int32 Wout =
+        static_cast<int>(ceil(static_cast<float>(((W + 2 * pad) - border * 2)) /
+                              static_cast<float>(stride_1)));
+
+    OP_REQUIRES(context, Hout >= 1,
+                errors::InvalidArgument(
+                    "Neighborhood and kernel don't fit in input height."));
+    OP_REQUIRES(context, Wout >= 1,
+                errors::InvalidArgument(
+                    "Neighborhood and kernel don't fit in input width."));
+
+    Tensor* output_t;
+    OP_REQUIRES_OK(
+        context, context->allocate_output(0, TensorShape({N, Cout, Hout, Wout}),
+                                          &output_t));
+
+    Tensor padded_a_t;
+    Tensor padded_b_t;
+    TensorShape padded_shape({N, H + 2 * pad, W + 2 * pad, C});
+    OP_REQUIRES_OK(context, context->allocate_temp(DataTypeToEnum<T>::value,
```

Comment:
This is unused in the CPU version. Should I sub-class "CorrelationCostOpGPU : public CorrelationCostOp"?